File: implementation/Python/synapses_py/model/encoding/discrete_attribute.py
from typing import Dict

from functional import seq
from functional.pipeline import Sequence

from synapses_py.model import utilities
from synapses_py.model.encoding.serialization import DiscreteAttribute, DiscreteAttributeSerialized


def updated(datapoint: Dict[str, str],
            discrete_attribute: DiscreteAttribute
            ) -> DiscreteAttribute:
    exist = discrete_attribute \
        .values \
        .exists(lambda x: x == datapoint[discrete_attribute.key])
    updated_vals = discrete_attribute.values if exist else \
        utilities.lazy_cons(
            datapoint[discrete_attribute.key],
            discrete_attribute.values
        )
    return DiscreteAttribute(discrete_attribute.key, updated_vals)


def encode(value: str,
           discrete_attribute: DiscreteAttribute
           ) -> Sequence:
    return discrete_attribute \
        .values \
        .map(lambda x: 1.0 if x == value else 0.0)


def decode(encoded_values: Sequence,
           discrete_attribute: DiscreteAttribute
           ) -> str:
    return discrete_attribute \
        .values \
        .zip(encoded_values) \
        .reduce(lambda acc, x:
                x if x[1] > acc[1] else
                acc)[0]


def serialized(discrete_attribute: DiscreteAttribute
               ) -> DiscreteAttributeSerialized:
    return DiscreteAttributeSerialized(
        discrete_attribute.key,
        discrete_attribute.values.to_list()
    )


def deserialized(discrete_attribute_serialized: Dict
                 ) -> DiscreteAttribute:
    return DiscreteAttribute(
        discrete_attribute_serialized['key'],
        seq(discrete_attribute_serialized['values'])
    )

File: implementation/Scala/src/main/scala/synapses/model/encoding/DiscreteAttribute.scala
package synapses.model.encoding

import synapses.model.encoding.Serialization._

object DiscreteAttribute {

  def updated(datapoint: Map[String, String])
             (discreteAttribute: DiscreteAttribute)
  : DiscreteAttribute = {
    val exists = discreteAttribute
      .values
      .contains(datapoint(discreteAttribute.key))
    val updatedValues =
      if (exists) discreteAttribute.values
      else LazyList.cons(
        datapoint(discreteAttribute.key), // TODO: fetch `datapoint(discreteAttribute.key)` only once
        discreteAttribute.values
      )
    discreteAttribute.copy(values = updatedValues)
  }

  def encode(value: String)
            (discreteAttribute: DiscreteAttribute)
  : LazyList[Double] =
    discreteAttribute
      .values
      .map(s => if (s == value) 1.0 else 0.0)

  private def decodeAccF(acc: (String, Double),
                 x: (String, Double))
  : (String, Double) =
    if (x._2 > acc._2) x
    else acc

  def decode(encodedValues: LazyList[Double])
            (discreteAttribute: DiscreteAttribute)
  : String = discreteAttribute
    .values
    .zip(encodedValues)
    .reduceLeft(decodeAccF)
    ._1

  def serialized(discreteAttribute: DiscreteAttribute)
  : DiscreteAttributeSerialized = DiscreteAttributeSerialized(
    discreteAttribute.key,
    discreteAttribute.values.toList
  )

  def deserialized(discreteAttributeSerialized: DiscreteAttributeSerialized)
  : DiscreteAttribute = Serialization.DiscreteAttribute(
    discreteAttributeSerialized.key,
    discreteAttributeSerialized.values.to(LazyList)
  )

}
